File: Pattern.java

//calculates the N-th term
//for a given N

import java.util.*;
import java.lang.*;

public class Pattern{
	static List<Integer> user_input = new ArrayList<Integer>();
	static List<Integer> prime_list = new ArrayList<Integer>();
	static List<Integer> solution_arrayList = new ArrayList<Integer>();


	//generate prime numbers
	//in the range; min -max user input
	static void generatePrime(){
      int i =0;
      int num =0;

      int last_value = user_input.get(user_input.size()-1); // last value
      int first_element = user_input.get(0);
      for (i = first_element; i <= last_value; i++)
      {
         int counter=0;
         for(num =i; num>=1; num--)
         {
	    if(i%num==0)
	    {
		counter +=1;
	    }
	 }
	 if (counter ==2)
	 {

	    prime_list.add(i);
	 }
      }
   }

//calculate the nearest 
// to current value of user input
 
	
public static void series() {

   for(int key:user_input){
    int index = Collections.binarySearch(prime_list, key);
    int closest;
    int nTh_term =0;
    if (index >= 0) {
        closest = prime_list.get(index);
        nTh_term = Math.abs(key - closest);
        solution_arrayList.add(nTh_term);
    } else {
        index = -index - 1;
        if (index == 0){
            closest = prime_list.get(index);
            nTh_term = Math.abs(key - closest);
        	solution_arrayList.add(nTh_term);
            
        } else if (index == prime_list.size()){
            closest = prime_list.get(index - 1);
            nTh_term = Math.abs(key - closest);
        	solution_arrayList.add(nTh_term);
        } else {
             
            int prev = prime_list.get(index - 1);
            int next = prime_list.get(index);
            
            int prev_diff = Math.abs(key - prev);
            int next_diff = Math.abs(key-next);
            
            if(prev_diff > next_diff){
                closest = next;
                solution_arrayList.add(next_diff);
            }
          
            else if (prev_diff<=next_diff){
                closest = prev;
                solution_arrayList.add(prev_diff);
            }
            else{
                closest = prev;
                solution_arrayList.add(prev_diff);
            }
            
            
            
        }
    }
} 
}




	public static void main(String[] args){
		Scanner scan = new Scanner(System.in);
		System.out.println("Input:");
		int test_case = scan.nextInt();
		int place_holder = 0;


		if(1<=test_case || test_case <=200){
		for(int i=0;i<test_case;i++){
			//System.out.println("value:");
			int input_value = scan.nextInt();
			if(input_value>=1 || input_value<=1000000){
				place_holder = input_value;
				user_input.add(place_holder);
				place_holder = 0;

		}

	}

	}

	Collections.sort(user_input); // Sorts list in ascending order
	generatePrime(); //generate all posible prime numbers in the range [N,10^6] / [first_prime,10^6] - if N < first_prime
	series();
	
	System.out.println();
	System.out.println("Output");
	for(int g:solution_arrayList){
		System.out.println(g);
	}

		}
	}

File: Pattern.php


<?php
//calculates the N-th term
//for a given N

    class Pattern{
      protected static $user_input; //int []
      protected static $prime_list; // int []


      public static function __staticinit(){
        //static class members
        self::$user_input=  array();
        self::$prime_list= array();
      }


      //generate prime numbers
	     //in the range; min -max user input
      protected static function getPrime(){

        $i=0;
        $num=0;
        $prime_index =0;
        self::$prime_list = array();
        $last_value = self::$user_input[(count(self::$user_input))-1];
        $first_value = self::$user_input[0];
        self::$prime_list = array();

        for($i=$first_value;$i<=$last_value;$i++){

          $counter = 0;
          for($j=1;$j<=$i;$j++){


                if($i % $j==0){

                      $counter++;
                }
          }


        if($counter==2){

               array_push(self::$prime_list,$i);
             }
          }

      }

      // binary search function
      function binarySearch(Array $sortedPrimeList, $key){

          if (count($sortedPrimeList) === 0){
              $value = 0;
              }
          $low = 0;
          $high = count($sortedPrimeList) - 1;

          while ($low <= $high) {


              $mid = floor(($low + $high) / 2);


                if($sortedPrimeList[$mid] == $key) {
                    return $mid; // middle value
                  }

                if ($key < $sortedPrimeList[$mid]) {

                    $high = $mid -1; // search on the left
                  }
                else {

                  $low = $mid + 1; // on the right
                  }

                }
              return -($low + 1); // if not found
    }


      //calculate the nearest
      // to current value of user input

      public static function series(){
        foreach(self::$user_input as $key){
          $index = Pattern::binarySearch(self::$prime_list,$key);
          $closest = 0;
          $nTh_term = 0;

          if(($index >= 0)){
            $closest = self::$prime_list[$index];
            $nTh_term = abs(($key - $closest));
            print($nTh_term."\n");

          }
          else{
            $index = (-$index -1);
            if(($index == 0)){
              $closest = self::$prime_list[$index];
              $nTh_term = abs(($key - $closest));
              print($nTh_term."\n");


            }
            elseif (($index==count(self::$prime_list))) {
              $closest = self::$prime_list[($index - 1)];
              $nTh_term = abs(($key - $closest));
              print($nTh_term."\n");

            }
            else{
              $prev = self::$prime_list[($index -1)];
              $next = self::$prime_list[$index];


              $prev_diff = abs(($key - $prev));
              $next_diff = abs(($key - $next));

              if(($prev_diff > $next_diff)){
                $closest = $next;
                print($next_diff."\n");
              }
              elseif (($prev_diff <= $next_diff)) {
                $closest = $prev;
                print($prev_diff."\n");
              }
              else{
                $closest = $prev;
                print($prev_diff."\n");
              }
            }
          }
        }

      }

      // equivalent java main method
      public static function main(){
        print("Input:"."\n");
        $test_case = rtrim(fgets(STDIN));
        if((($test_case >=1) ||($test_case <=200))){
          self::$user_input = array();
          for($i=0;($i<$test_case);++$i){
            $input_value = rtrim(fgets(STDIN));
            if((($input_value >=1) || ($input_value <=1000000))){
              array_push(self::$user_input,$input_value);
          }
        }
      }
      sort(self::$user_input); // sorts array in ascending order
      Pattern::getPrime();
      print(""."\n");
      print("Output:"."\n");
      Pattern::series();

    }
  }
  Pattern::main();
  Pattern::__staticinit(); //initialize static variables on load





 ?>
